TST: Add test for dual_annealing accept_reject

sethtroisi opened the pull request:
#### Reference issue
<!--Example: Closes gh-WXYZ.-->
TST to demonstrate #10892

#### What does this implement/fix?
<!--Please explain your changes.-->

#### Additional information
<!--Any additional information you think is important.-->

I'm second guessing my math so I added this test.

See notes in [scipy.optimize.dual_annealing](https://docs.scipy.org/doc/scipy/reference/generated/scipy.optimize.dual_annealing.html) or [DOI: 10.1103/PhysRevE.62.4473](https://www.researchgate.net/publication/12238832_Efficiency_of_generalized_simulated_annealing)
For what the accept rate should be.

sethtroisi commented on scipy/optimize/tests/test__dual_annealing.py:
```diff
@@ -258,3 +259,46 @@ def test_from_docstring(self):
+
+        print ("\te:", e, "rate:", rate, "accepted:", count_accepted)
+
+        if e < current_energy:
+            assert rate >= 1
+            assert count_accepted == iterations
+        else:
+            assert_allclose(rate, 0.9183859021684453)
+            assert count_accepted == 911
```
Currently
```
    e: 0 rate: 1.0576809264052165 accepted: 848
    e: 2 rate: 0.9183859021684453 accepted: 0
```

After we move the +1 outside of a paren in #10892 this test passes

sethtroisi commented:
@rgommers This is nice to prove that current calculation in broken (and that #12266 fixes the calculation)

sethtroisi commented on scipy/optimize/tests/test__dual_annealing.py:
```diff
@@ -258,3 +259,46 @@ def test_from_docstring(self):
+        pqv = 1 - (1 - accept_param) * (e - current_energy) / temp_step
+        rate = np.exp(np.log(pqv) / (1 - accept_param))
+
+        print("accept_reject e:", e, "rate:", rate, "accepted:", count_accepted)
```
This will be removed when I commit #10892

sethtroisi commented:
@sgubianpm and @WarrenWeckesser This test demonstrates how accepted != accept_rate. After this tests goes in it's as simple as moving a parenthesis in #10892

rgommers commented:
Thanks @sethtroisi. Sorry I'm a bit unresponsive at the moment, I'm on holiday and have a bit of a backlog.

Is there a reason why you split the regression test and the fix into separate PRs? Also, this isn't `xfail`ed at the moment - if it's a regression test, it should fail here right?

sethtroisi commented:
@rgommers Have fun on vacation (I noticed you were out which is why I added warren).

I split out the TST to prove that this is broken. If you want you can submit #12266 directly.

I initially had this marked as `@xfail` but that hides my own errors (I had several), so I left a `# TODO gh-10892 Fix incorrect accepted counts.` which is fixed in #12266 (see https://github.com/scipy/scipy/pull/12266/commits/046898e4cf0bb4b70f50dbd2fb1829c304c47f28)

sethtroisi commented:
@rgommers Sorry to keep pinging on this.

To summarize:

1. [Reference 4](https://www.researchgate.net/publication/12238832_Efficiency_of_generalized_simulated_annealing/link/542ad4a60cf277d58e8a09f7/download) at  [scipy.optimize.dual_annealing](https://docs.scipy.org/doc/scipy/reference/generated/scipy.optimize.dual_annealing.html) says
    > The jump is accepted if it is downhill (in terms of the energy function). If
the jump is uphill it might be accepted according to an acceptance probability

    This also matches the formula as given in the scipy documentation `(1 - (1 - qa) * B * deltaE)^(1/(1-qa))`
with `ΔE = current energy - best energy`, `B = 1 / K * initial Temp`
`(1 - (1- -5) * (1/5230) * (5 - 10)) ^ (1/(1- -5)) = 1.0009`, accept prob = `min(1, 1.00009) = 1` = 100% of time

    This PR shows the behavior of `accept_reject` is bad, only 633 of 1000 iterations accept when `new_energy=0` and `current_energy=1` (see parameterized test `(0, 100, 633, 1.0097587941791923)`) and never accepts with higher energy even when temperature is varied.

2. This is clearly caused by a bad parenthesis which is fixed in #12266. #12266 fixes the parenthesis and updates these tests so that the third column (observed accepts) matches the calculated accept_rate (4th column).

    1. The original author sgubianpm commented in #10892 that this is a typo and should be fixed.

3. Technically these tests could be marked `@xfail` in this PR but I'm not a perfect programmer and marking `@xfail` suppresses other failures (bad variable names, missed calculations) so I preferred the comment here `# TODO gh-10892 Fix incorrect accepted counts` because both PRs will be submitted back to back and this quickly becomes a normal test.

rgommers commented:
Thanks @sethtroisi. Let me close this, since gh-12266 is a superset of this PR (same test, but with the actual bug fix).

The clear summary is helpful, thanks again.
